flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/KeyedSerializationSchema.java

```diff
@@ -55,4 +57,13 @@
 	 * @return null or the target topic
 	 */
 	String getTargetTopic(T element);
+
+	/**
+	 *
+	 * @param element The incoming element to be serialized
+	 * @return collection of headers (maybe empty)
+	 */
+	default Iterable<Map.Entry<String, byte[]>> headers(T element) {
```

Comment:
I would even suggest to rewrap Kafka Header class into Flink class here, because it should be more extensible if Kafka decides to extend its Header in future. But if it is possible to accept just Kafka `ConsumerRecord` directly for all versions then we do not need any `Record` wrapping at all.

Reply:
Not sure how calling one constructor can be easier than calling another... Map.Entry is more convenient in case headers representation in element is Map (or Guava's Multimap), in that case it just particular implementation would be element.entrySet()

Reply:
I can try to use ProducerRecord by adding factory method to KeyedSerializationSchema
```suggestion
interface KeyedSerializationSchema {
...
	default ProducerRecord<byte[], byte[]> newProducerRecord(T element, 
	byte[] serializedKey, byte[] serializedValue, Integer partition, String topic, Long timestamp) {
		return new ProducerRecord<>(topic, partition, timestamp, serializedKey, serializedValue);
	}
}
```
And then in a producer:
```suggestion
	String targetTopic = schema.getTargetTopic(next);
	if (targetTopic == null) {
		targetTopic = defaultTopicId;
	}
	...
	ProducerRecord<byte[], byte[]> record = schema.newProducerRecord(
	  next, 
	  serializedKey,
	  serializedValue
	  flinkKafkaPartitioner != null ?
		flinkKafkaPartitioner.partition(next, serializedKey, serializedValue, targetTopic, partitions) : null,
	  targetTopic, 
	  timestamp);
	  ...
```
Note that we can't deprecate getTargetTopic, serializeKey and serializedValue, because they are used by partitioner as well.

What do you think?

Reply:
Can't have default method creating `ProducerRecord`, because 0.8 doesn't have timestamp parameter in ProducerRecord constructor

Reply:
@azagrebin, so you suggest to have Flink class to represent Kafka Header instead of Map.Entry?

Reply:
@alexeyt820 
Ideally we could deprecate the partitioner in the producer constructer as well because ProducerRecord already contains partition which user can assign for the record.

Instead of deprecating the methods in (de)ser schema interfaces we could deprecate them fully and introduce new `Kafka(De)SerializationSchema` interfaces which work with Kafka `Consumer/ProducerRecord` classes.

We also introduce adaptors from older schemas to the newer ones. The producer/consumer constructors, which currently accept older schemas, will use adaptors to create newer schemas. The actual code should also work with newer schemas only.

The serialization schema adaptor could optionally take the `FlinkKafkaPartitioner` to populate the partition in similar way as now it happens directly in producer.

What do you think?

Reply:
Let's say we have 
```
interface KafkaSerializationSchema<T> {
  ProducerRecord toKafka(T element)
```
Adapter for KafkaSerializationSchema will be Kafka version specific, so we will need to have `KafkaSerializationSchemaAdapter08<T>`, `KafkaSerializationSchemaAdapter09<T>`.
Also if we replacing partitioner we need to take into account, that the partitioner had initialization method - open and extract arguments, so it will become something like that
```
interface KafkaSerializationSchema<T> {
  void open(int parallelInstanceId, int parallelInstances);
  ProducerRecord toKafka(T element, int[] partitions);
```
in my opinion it doesn't look clean. 

Alternative, we can keep Partitioner as is, but introduce version specific `ProducerRecord reparation( ProducerRecord source, int partition)` in _FlinkKafkaProduceXY_.
Then FlinkKafkaProducer can do following:
```
  ProducerRecord record = schema.toKafka(value);
  if (flinkKafkaPartitioner != null) {
   int partition = flinkKafkaPartitioner.partition(value, record.key(), record.value(), record.topic(), partitions);
  if(record.partition() == null || partition != record.partition()) {
    record = this.reparation(record,partition);
  }
```
So if we have `flinkKafkaPartitioner` and if it chooses different partition, we will have extra object allocation. Maybe not so bad?
What is your opinion?

Reply:
Let's say the new serialization schema looks like this:
```
interface KafkaSerializationSchema<T> {
  ProducerRecord serialize(T element, @Nullable Long timestamp, PartitionInfo partitionInfo);
}
class PartitionInfo {
  private final int parallelInstanceId, int parallelInstances, int[] partitions);
  // ....
}
```
This way it should preserve previous Partitioner API capabilities.

Regarding the multiple adaptors, from what I see, the difference is that the older connectors do not support timestamp in `ProducerRecord`. Could we have one adaptor in base module which calls `ProducerRecord` constructor without timestamp if `KafkaSerializationSchema.serialize` is called with `timestamp = null`? Producers, which do not supported timestamp, could call new schema with `timestamp = null`, including base one (although `ProducerRecord` has `timestamp` in base).

Regarding the open method, could adaptor or partitioner still be opened the similar way as it happens now to current partitioner? I agree that the new schema would look weird with the open method.

`PartitionInfo` could be populated in a similar way as `int[] partitions` are cached now in `FlinkKafkaProducer.topicPartitionsMap` in `FlinkKafkaProducer.invoke`, just `parallelInstanceId` and `parallelInstances` are added.

Reply:
So we still need two KafkaSerializationSchema adapters: one for base, 0.8 and 0.9 which created ProducerRecord w/o timestamp and one with timestamp for rest of versions?

Regarding `open`, I guess we can do it, but if someone had non-trivial overridden `open` and want to migrate to new `KafkaSerializationSchema` it will be more challenging, because he will need to move logic of `open`, which was called once into `serialize`, which is called for each record. Likely rare case, and it is feasible to cache results of doing "open" during `serialize` in `KafkaSerializationSchema`. So let's consider it acceptable?

Reply:
Yes, I would try with 2 adaptors and accept the migration risk. API w/o open method looks simpler. It can be extended later anyways if we discover use cases for it.

Reply:
What about `defaultTopic`? We need it, because `KeyedSerializationSchema.getTargetTopic` may (and I guess usually does) return null. We can add `defaultTopic` as parameter to `KafkaSerializationSchema.serialize` or we can add it to `PartitionInfo`. 

And `int[] partitions` seems should be method `int[] PartitionInfo.partitions(String topicName), because `KeyedSerializationSchema.serialize` need ability to query partitions for arbitrary topic.

Reply:
Also we can provide `defaultTopic` via constructor to an adapter. And for assume similar approach for other `KeyedSerializationSchema` implementations.

Reply:
True, it should be `PartitionInfo.getPartitions(topic)` and the adaptor has to have the `defaultTopic`.
Maybe, we could pack `topicPartitionsMap` and `getPartitionsByTopic` logic into `PartitionInfo` which would be one object belonging to `FlinkKafkaProducer`.

Reply:
Or may be it should be `ProducerRecord serialize(T element, String defaultTopic, @Nullable Long timestamp, PartitionInfo partitionInfo);`, otherwise every implementing class will need to have `topic` as constructor argument.

Reply:
I would keep it in the constructor because it will be the same for each `serialize` method. We could consider having overloaded `PartitionInfo.getPartitions()` method but this would also mean some implicit semantics of `PartitionInfo` having `defaultTopic` from `FlinkKafkaProducer`.

Reply:
4db60d2

Reply:
Thanks, I will have a look. 

Do you also plan to introduce the similar `KafkaDeserialisationSchema` to covert `ConsumerRecord` into object to have symmetrical API?

Also, it would be nice to split this big PR into two PRs with separate consumer and producer refactoring. The consumer part seems to be more clear and popular as a feature, we could try to merge it first.

Reply:
I didn't plan to introduce `KafkaDeserialisationSchema`, current approach with default method is simpler, but if you think that it is important to have symmetrical API, I can do it. Let me know.

I don't like idea of splitting into two PRs, currently added tests produce records with headers and consumes them. I feel splitting will add extra complexity at least from unit test standpoint. And for my purposes I need both - consumer and producer with headers.

Reply:
The change for `KafkaDeserialisationSchema` might be simple if we do it like this:
```
// new API, that we could now use everywhere in Kafka code instead of KeyedDeserializationSchema
interface KafkaDeserialisationSchema<T> {
  T deserialize(ConsumerRecord<byte[], byte[]> record);
}

// old API, stays almost the same as we already have it in the PR and also supports the new API
@deprecated
interface KeyedDeserializationSchema<T> extends KafkaDeserialisationSchema<T> {
  T deserialize(byte[] messageKey, byte[] message, String topic, int partition, long offset);
  default T deserialize(ConsumerRecord<byte[], byte[]> record) {
    return deserialize(record.key(), record.value(), record.topic(), record.partition(), record.offset());
  }
}
```
Do you think it is quick enough change?

Ok, I see your point with tests regarding the splitting. The thing is that we are interested in merging this PR this week before feature freeze for 1.8 release (planned this Friday). Let's see then how it goes and how much we can manage. Do you have time this week to work on this?

Reply:
It is quick enough. Keep in mind that it also affects all constructors etc, so it is more then 2 files change. And there were conflicting changes in `KafkaConsumerBase` and `KafkaConsumerBaseTest` so had to merge upstream master in PR branch.
Hopefully 7e6753b will be ok